Sources/NIOPerformanceTester/main.swift

```diff
@@ -0,0 +1,504 @@
+    private let isDonePromise: EventLoopPromise<Int>
+    
+    init(numberOfRequests: Int, eventLoop: EventLoop) {
+        self.remainingNumberOfRequests = numberOfRequests
+        self.numberOfRequests = numberOfRequests
+        self.isDonePromise = eventLoop.newPromise()
+    }
+    
+    func wait() throws -> Int {
+        let reqs = try self.isDonePromise.futureResult.wait()
+        precondition(reqs == self.numberOfRequests)
+        return reqs
+    }
+    
+    func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        ctx.channel.close(promise: nil)
+        self.isDonePromise.fail(error: error)
+    }
+    
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        let reqPart = self.unwrapInboundIn(data)
+        if case .end(nil) = reqPart {
+            if self.remainingNumberOfRequests <= 0 {
+                ctx.channel.close().map { self.doneRequests }.cascade(promise: self.isDonePromise)
+            } else {
+                self.doneRequests += 1
+                self.remainingNumberOfRequests -= 1
+                
+                ctx.write(self.wrapOutboundOut(.head(head)), promise: nil)
+                ctx.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+            }
+        }
+    }
+}
+
+// MARK: Performance Tests
+
+measureAndPrint(desc: "bytebuffer_lots_of_rw") {
+    let dispatchData = ("A" as StaticString).withUTF8Buffer { ptr in
+        DispatchData(bytes: UnsafeRawBufferPointer(start: UnsafeRawPointer(ptr.baseAddress), count: ptr.count))
+    }
+    var buffer = ByteBufferAllocator().buffer(capacity: 7 * 1024 * 1024)
+    let foundationData = "A".data(using: .utf8)!
+    @inline(never)
+    func doWrites(buffer: inout ByteBuffer) {
+        /* all of those should be 0 allocations */
+
+        // buffer.write(bytes: foundationData) // see SR-7542
+        buffer.write(bytes: [0x41])
+        buffer.write(bytes: dispatchData)
+        buffer.write(bytes: "A".utf8)
+        buffer.write(string: "A")
+        buffer.write(staticString: "A")
+        buffer.write(integer: 0x41, as: UInt8.self)
+    }
+    @inline(never)
+    func doReads(buffer: inout ByteBuffer) {
+        /* these ones are zero allocations */
+        let val = buffer.readInteger(as: UInt8.self)
+        precondition(0x41 == val, "\(val!)")
+        var slice = buffer.readSlice(length: 1)
+        let sliceVal = slice!.readInteger(as: UInt8.self)
+        precondition(0x41 == sliceVal, "\(sliceVal!)")
+        buffer.withUnsafeReadableBytes { ptr in
+            precondition(ptr[0] == 0x41)
+        }
+
+        /* those down here should be one allocation each */
+        let arr = buffer.readBytes(length: 1)
+        precondition([0x41] == arr!, "\(arr!)")
+        let str = buffer.readString(length: 1)
+        precondition("A" == str, "\(str!)")
+    }
+    for _ in 0 ..< 1024*1024 {
+        doWrites(buffer: &buffer)
+        doReads(buffer: &buffer)
+    }
+    return buffer.readableBytes
+}
```

Comment:
This test is identical to the allocation one: is there a way we can share the code, rather than duplicate it?

Reply:
@Lukasa it's not exactly identical (IIRC 1000 runs vs. 1024*1024 runs). Sharing code there seems pretty hard unfortunately :|